src/Features/CSharp/Portable/ExtractMethod/CSharpMethodExtractor.Analyzer.cs

```diff
@@ -129,6 +131,54 @@ protected override bool ReadOnlyFieldAllowed()
                 var scope = this.SelectionResult.GetContainingScopeOf<ConstructorDeclarationSyntax>();
                 return scope == null;
             }
+
+            protected override ITypeSymbol GetSymbolType(SemanticModel semanticModel, ISymbol symbol)
+            {
+                var selectionOperation = semanticModel.GetOperation(this.SelectionResult.GetContainingScope());
+
+                switch (symbol)
+                {
+                    case ILocalSymbol localSymbol when localSymbol.NullableAnnotation == NullableAnnotation.Annotated:
+                    case IParameterSymbol parameterSymbol when parameterSymbol.NullableAnnotation == NullableAnnotation.Annotated:
+
+                        // For local symbols and parameters, we can check what the flow state 
+                        // for refences to the symbols are and determine if we can change 
+                        // the nullability to a less permissive state.
+                        var references = selectionOperation.DescendantsAndSelf()
+                            .Where(IsSymbolReferencedByOperation);
+
+                        var flowState = GetFlowStateFromLocations(references);
+                        return base.GetSymbolType(semanticModel, symbol).WithNullability(flowState);
+
+                    default:
+                        return base.GetSymbolType(semanticModel, symbol);
+                }
+
+                NullableFlowState GetFlowStateFromLocations(IEnumerable<IOperation> references)
+                {
+                    foreach (var reference in references)
+                    {
+                        var typeInfo = semanticModel.GetTypeInfo(reference.Syntax);
+
+                        switch (typeInfo.Nullability.FlowState)
```

Comment:
Does it matter if the locations have a mix of None and MaybeNull? Which one has to be returned then?

Reply:
It shouldn't matter, but also a good test case to have a few #enable and #disables in the method

Reply:
I wasn't sure if it's worth simplifying the method then to just be Boolean-returning since if the difference doesn't matter this otherwise looks like a bug.

Reply:
Sure, changed to `AreAllReferencesNotNull`